### ISSUE TITLE
Adapt messages received by Behavior (contramap)

### ISSUE THREAD

**Issue by TomasMikula** [akka/akka#26321]
### What

I propose to add a method to `Behavior` that would allow to adapt the message before it is received by the behavior.

```scala
class Behavior[T] {
  def contramap[U](f: U => T): Behavior[U]
}
```

### Related

There is a method on `ActorContext`, called [messageAdapter](https://doc.akka.io/api/akka/current/akka/actor/typed/scaladsl/ActorContext.html#messageAdapter[U](f:U=%3ET)(implicitevidence$1:scala.reflect.ClassTag[U]):akka.actor.typed.ActorRef[U]), which serves a similar purpose:

```scala
trait ActorContext[T] {
   def messageAdapter[U](f: (U) ⇒ T)(implicit arg0: ClassTag[U]): ActorRef[U] 
}
```

However, the benefit of the method I propose is that messages can be adapted before any actor is spawned.

### Use case

An actor that sends messages (of type `SelfMsg`) to itself, but hides this fact from the outside world: the outside world is only able to send messages of type `FooMsg` (unrelated to `SelfMsg`) to that actor.

The solution to this problem would be to define a `Behavior` accepting `Either[SelfMsg, FooMsg]` and then adapt it to only receive `FooMsg`:

```scala
private def privateBehavior: Behavior[Either[SelfMsg, FooMsg]] =
  Behaviors.withTimers { scheduler =>
    // can use scheduler to send Left(SelfMsg) to self
    ???
  }

def behavior: Behavior[FooMsg] = privateBehavior.contramap(Right(_))
```

An actor accepting `Either[SelfMsg, FooMsg]` would never be created (or at least would not be accessible to anyone).

**Comment by johanandren** [comment]
Doing the contra map thing is already covered by `Behaviors.intercept` (although a little bit more verbose).

For the internal message protocol, a better way to than contra mapping is to have private internal subtypes of the message, which hides the sub type from external users, this also avoids the overhead of allocating an extra wrapper for every incoming message:

```scala
object MyBehavior {
  sealed trait MyMessages
  case class PublicMessage() extends MyMessages
  private sealed trait InternalMessages extends MyMessages
  // or internal messages without a common supertype
  private case class MyInternalMessage() extends MyMessage
}
```

Also, note that in general having message protocols with type parameters will be a pain since it cannot be matched on at runtime.

**Comment by TomasMikula** [comment]
Thanks for the quick response!

The public message is potentially (and actually, in my concrete case) defined somewhere else, so I can't make it extend some superclass. An extra wrapper (or translation to a new type) is therefore unavoidable.

Thanks for pointing me to `Behaviors.intercept`, that will do for the time being.

> Also, note that in general having message protocols with type parameters will be a pain since it cannot be matched on at runtime.

In my example, `FooMsg`, `SelfMsg` (and therefore `Either[SelfMsg, FooMsg]`) were concrete types (note that my `def privateBehavior` and `def behavior` do not take any type parameters). Does your comment still apply? I don't foresee any problem in such a case.

**Comment by TomasMikula** [comment]
Yeah, the extra allocation will be there one way or another. The actor is consuming messages produced by someone else. I can either
 - wrap it (into `Either` or similar); or
 - define a custom message type alongside the actor and then translate the externally produced message into this custom message. Again, this is an extra allocation.

**Comment by johanandren** [comment]
So it is more of a subscription/response than a part of the actor protocol? 

Then I'd recommend going for the internal protocol wrapper message and the adater - it's what it is there for (note that the overhead in this case is just for those messages vs all messages with the `Either` solution). 

If that alloc overhead is too large because the actor is a hot spot you could also do tricks (tm) such as internally defining the actor to accept a wider type (worst case `Any`) but then `.narrow` it down before returning it from a factory method to hide that from a public API. That trick can complicate things if someone else expects to be able to intercept messages, the interceptor all of a sudden will see messages of the wrong type, so you may have to filter those.

**Comment by johanandren** [comment]
The message adapters should be the goto-solution for this kind of thing, and for a case were you want to work around the allocation I think it is ok if that is a tiny bit more cumbersome, or else it will encourage people to go down that path by default, so let's close this for now. Thanks for the discussion around it!

**Comment by johanandren** [comment]
The interception error happens when your actor are receiving other kinds of messages than it stated where you added the interceptor, you could either do a broader interception on the incoming side, or override the `interceptMessageType` to be selective about what types of messages to intercept.

For the adapter, the actor controls its own lifecycle, so if it should subscribe somewhere from start, setup is a good place, if it should happen later, as a reaction to a message that's also reasonable. Since you can compose the behaviors, the core behavior does not even need to know this.

```scala
object MyBehavior {
  sealed trait Message
  final case class Command(...) extends Message
  private final case class Adapted(msg: External) extends Message

  def behavior(publisher: ActorRef[Subscribe]): Behavior[Message] = Behaviors.setup { ctx => 
    val adapter = ctx.messageAdapter(Try[External] => Adapted) 
    publisher ! Subscribe(adapter)

    Behaviors.receiveMessage {
      case Command(...) => ...
      case Adapted(msg) => ...
   }
}
```

**Comment by TomasMikula** [comment]
> The interception error happens when your actor are receiving other kinds of messages than it stated

The fact that a typed actor might legally receive other types of messages than it stated is most surprising. But, obviously, the test kit sends such messages.

`interceptMessageType` helped, thanks for pointing me to it!

---

The setup of the actors in my case is the other way around compared to your code snippet. It is not that my `Behavior` subscribes to a producer whose `ActorRef` it receives as a parameter. Rather, the producer actor is initialized with consumer `ActorRef`s. That's why `messageAdapter` does not seem to be of any help.

```scala
object Producer {
  def behavior(consumers: ActorRef[External]): Behavior[Foo]
}

object MyBehavior {
  private def internalBehavior: Behavior[Message] = { /* ... */ }
  def behavior: Behavior[External] = ??? // messageAdapter does not help to implement this
}
```

**Comment by johanandren** [comment]
> The fact that a typed actor might legally receive other types of messages than it stated is most surprising

Yeah it is a bit tricky and one reason we'd generally recommend adapters instead of  `Behavior[SuperType].narrow[SubType]`. We could have avoided having those passed to the interceptor in the face of erasure at the cost of a classtag everywhere and we considered that too high a tax to pay (especially for the Java side of the API where users would have to provide the class by hand).

Since we don't have union types (to express heterogeneous protocols like `ActorRef[Message | External]`) the only solution I can think of is to go for least common supertype (probably `AnyRef`)  and directly `narrow` down the actorref to the two sub-protocols in one place and pass those refs around.

**Comment by johanandren** [comment]
I have thought some more about this, and now I'm starting to think that maybe there is a missing piece of functionality here, to allow contramap on the `ActorRef` rather than the behavior. If someone gives you a readymade `ActorRef[A]` and you need a `ActorRef[B]` and have access to a `A => B`. I wonder how/where such a transformation would execute though, caller thread?

**Comment by johanandren** [comment]
One complication/showstopper with contramap on actorref is that you must be able to send `ActorRef`s to another node, and they therefore cannot contain a function (not serializable), so it'd have to be an actual actor doing the transformation (which you can already do today, by just spawning an intermediate actor).

**Comment by TomasMikula** [comment]
> > The fact that a typed actor might legally receive other types of messages than it stated is most surprising
> 
> Yeah it is a bit tricky and one reason we'd generally recommend adapters instead of `Behavior[SuperType].narrow[SubType]`.

The problem occurs even without `.narrow`. I suspect it is due to order of interceptors being undefined?

I have my `internalBehavior: Behavior[I]` which I intercept to get `Behavior[O]`:

```scala
object MyActor {
  def internalBehavior: Behavior[I] = { /* ... */ }
  def behavior: Behavior[O] = Behaviors.intercept(myInterceptor)(internalBehavior)
}
```

When someone else uses `MyActor.behavior` and intercepts it

```scala
def theirBehavior: Behavior[X] = Behaviors.intercept(theirInterceptor)(MyActor.behavior)
```

and then spawns an actor, I would expect that a message sent to that actor would first pass through `theirInterceptor` and only then through `myInterceptor`. But it seems that `ActorTestKit` intercepts my already intercepted behavior, but then `myInterceptor` is invoked first, which results in a `ClassCastException`. This would be prevented if `theirInterceptor` was invoked first.

**Comment by TomasMikula** [comment]
Ah, thanks for clarification. That's tricky...

**Comment by patriknw** [comment]
Related to #26867
